### PR TITLE
Quasi-static Finite-Strain Quasi-incompressible Visco-elasticity

### PR DESCRIPTION
**Pull request by jppelteret**
In this example, we extend `step-44` to parallel and implement a
rate-dependent constitutive law. We demonstrate the material behaviour
by loading a thin strip with a hole in its centre in tension, and track
the displacement of several vertices over time.

### CODE REVIEW

**Comment by jppelteret** [comment]
The animations are a bit bigger that I'd hoped (`6.3Mb` and `8.3Mb`) so I'm going to see if I can reduce them somehow.

**Comment by jppelteret** [comment]
I reduced the size of the animations to `2.4Mb` and `3.1Mb` without significant loss of image quality. I think that this is now ready for any comments.

File: Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
+## Overview
+Many rubber-like materials are not only near-incompressible in nature, but also exhibit a viscoelastic response (within the tested load and time scales).
+In this example, we extend the near-incompressible rate-independent constitutive used in `step-44` (which implements three-field quasi-static quasi-incompressible finite elasticity) to one that displays rate-dependent behaviour.
```
Comment (bangerth) [inline Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md]: no backticks around step-44 -- our scripts will link to it just fine

File: Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+It may seem that there is a contradiction of terms here, so lets clarify by saying the the problem remains "quasi-static" in the sense that inertial terms remain insignificant, even though the material response itself is rate-dependent.
```
Comment (bangerth) [inline Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md]: ha, what an interesting distinction. I guess it means that the timescales of material/stress relaxation are much longer than those of elastic wave propagation?

File: Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+We've also taken the opportunity to extend the code first shown in `step-44` to parallel (the primary reason for this contribution), using `Metis` as a grid partitioner and `Trilnos` for linear algebra.
```
Comment (bangerth) [inline Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md]: same here for step-44

File: Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+As a motivation as to why one might still choose to use `Metis` over `p4est` as a grid partitioner, at this point in time it is not possible to use `hp` the finite-element in conjunction with the distributed grid, meaning that this code could not, for example, be readily extended to the application shown in
```
Comment (bangerth) [inline Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md]: maybe associated METIS with `p::s::Triangulation` and p4est with `p::d::Triangulation` as that distinction may not be clear to all readers?

Comment (bangerth) [inline Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md]: `hp` the -> the `hp`

File: Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+* electro-elasticity
+* thermo-elasticity.
+
+Here we implement another permutation (one that is not necessarily benchmarked elsewhere), simply for demonstration purposes. The basic problem configuration is summarised in the following image.
+
+![Problem geometry](./doc/geometry.png)
+
+A thin strip of material with a circular hole is (in `3d`) constrained in the `Z` direction and loaded in the direction of its long edge.
+In our implementation, this load is applied to the `+Y` surface and may either be displacement control (a `Dirichlet` condition) or a traction load (a `Neumann` boundary condition).
+Due to the symmetry of both the geometry and load, the problem can be simplified by modelling only an eighth of the geometry in `3d` or a quarter in `2d`.
```
Comment (bangerth) [inline Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md]: We typically use American English, would would use "modeling". I don't know what you chose for the rest of the text (BE or AE) -- just stay consistent.

File: Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+the appropriate changes in the `main()` function.
+
+
+## Recommended Literature
+* C. Miehe (1994), Aspects of the formulation and finite element implementation of large strain isotropic elasticity. International Journal for Numerical Methods in Engineering 37 , 12, 1981-2004. DOI: [10.1002/nme.1620371202](http://doi.org/10.1002/nme.1620371202);
+* G.A. Holzapfel (2001), Nonlinear Solid Mechanics. A Continuum Approach for Engineering, John Wiley & Sons. ISBN: [978-0-471-82319-3](http://eu.wiley.com/WileyCDA/WileyTitle/productCd-0471823198.html);
+* P. Wriggers (2008), Nonlinear finite element methods, Springer. DOI: [10.1007/978-3-540-71001-1](http://doi.org/10.1007/978-3-540-71001-1);
+* T.J.R. Hughes (2000), The Finite Element Method: Linear Static and Dynamic Finite Element Analysis, Dover. ISBN: [978-0486411811](http://store.doverpublications.com/0486411818.html)
+
+The derivation of the finite-element problem, namely the definition and linearisation of the residual and their subsequent discretisation are quite lengthy and involved.
```
Comment (bangerth) [inline Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md]: -> linearization, discretization, if you want

File: Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+To begin, here is a comparison of the initial grid for the `2d` version of the problem (mirrored about the x-axis and rotated `90` degrees anti-clockwise)
+
+![Initial grid](./doc/grid_2d-mirrored.png)
+
+and of the final, displaced grid after the load has been applied and the material is in a (near-completely) relaxed state.
+
+![Displaced grid](./doc/grid_2d-final_mirrored.png)
+
+These results, as well as those that follow, were produced using the following material properties:
+* The poisson ratio is `0.4995`
```
Comment (bangerth) [inline Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md]: -> Poisson

**Comment by bangerth** [comment]
Once merged, can you please also propose a patch to the deal.II changelog to announce this? Also maybe an email to the mailing list?

File: Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md (hunk 5 above)
Comment (jppelteret) [inline Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md]: Unfortunately you'll have to continue to spell check my US English _ad infinitum_ 😝 I can't fix what I don't know is broken!

**Comment by jppelteret** [comment]
Thanks for the review @bangerth! I wouldn't recommend looking over the 1000's of lines of undocumented code either :-) With the exception of [this function](https://github.com/dealii/code-gallery/pull/23/files#diff-e9b1ba6a54b8bd208ffec3e73ea7c818R2246) (which I'd appreciate a quick evaluation of), I'm pretty confident that there are no bugs lurking around. The goal of this function is to compute the return the spatial position of a set of vertices, which may or may not be on the edge of a MPI partition (and therefore computed by more than one process).

File: Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md (hunk 5 above)
Comment (bangerth) [inline Quasi_static_Finite_strain_Quasi_incompressible_ViscoElasticity/README.md]: I don't actually care which spelling you use, I just wanted to point it out. I'm fine if you want to stick with BE (our South African English).

**Comment by jppelteret** [comment]
I know you don’t! I remember our last discussion, and I continue to agree that we should use AE as standard like in the rest of the library. You’ll just have to keep pointing out my “disability” 😂

**Comment by davydden** [comment]
looked at the videos and pictures -- a very nice example!

**Comment by jppelteret** [comment]
@bangerth I've made the changes that you requested. I'll also add an entry to the changelog and make a news post as you've suggested.

**Comment by bangerth** [comment]
> I'm pretty confident that there are no bugs lurking around.

That, by the way, would be the first time any code has no bugs. It's 
generally a good strategy to always assume that there are bugs :-) (That 
is, philosophically, why we have all of these assertions.)

**Comment by bangerth** [comment]
For reference the function `Solid<dim>::compute_vertex_positions` you asked me to review looks correct. It has no documentation, so I had to try to understand the code, but I think it looks reasonable if maybe a bit inefficient with the repeated MPI sums over points and over components. But unless you have evidence that that's too slow, there is no reason to optimize this.

**Comment by jppelteret** [comment]
> That, by the way, would be the first time any code has no bugs.

Ouch. That doesn't quite capture the spirit of what I meant! Lets increase the precision of my previous statement: 

> I'm pretty confident that there are no bugs lurking around.

Within a reasonable range of experimental parameters tested within the allocated time frame using the available computational resources, up to this point in time and to the best of my knowledge and recollection no unexpected behaviour has been observed when executing or post-processing the referenced source code.

I think that was what I meant to say :-)

**Comment by jppelteret** [comment]
> For reference the function `Solid<dim>::compute_vertex_positions` you asked me to review looks correct. It has no documentation, so I had to try to understand the code, but I think it looks reasonable if maybe a bit inefficient with the repeated MPI sums over points and over components. But unless you have evidence that that's too slow, there is no reason to optimize this.

Thank you! It wasn't documented at all so it was indeed unfair of me to ask. 

This function just tracks 4 points so its OK in terms of efficiency for now. But I'll apply your suggestion and optimise (and document) it a bit more over the weekend.

**Comment by bangerth** [comment]
On 02/24/2017 01:28 PM, Jean-Paul Pelteret wrote:
> Within a reasonable range of experimental parameters tested within
> the allocated time frame using the available computational resources,
> up to this point in time and to the best of my knowledge and
> recollection no unexpected behaviour has been observed when executing
> or post-processing the referenced source code.

Lawyered your way out of this hole here, didntcha? ;-)

**Comment by bangerth** [comment]
On 02/24/2017 01:34 PM, Jean-Paul Pelteret wrote:
> But I'll apply your suggestion and optimise (and document) it a bit more
> over the weekend.

No worries, not necessary. It was just something that occurred to me, 
but it's not worth doing much for just four points.

**Pull request merged**